### ISSUE TITLE
Postgresql auth reconnect problem

### ISSUE THREAD

**Issue by yshadow** [erlio/vernemq#609]
### Environment
Dockerized debian:jessie, deb package https://bintray.com/artifact/download/erlio/vernemq/deb/jessie/vernemq_1.3.0-1_amd64.deb

- VerneMQ Version: 1.3.0
- OS: debian jessie
- Erlang/OTP version (if building from source):

### Expected behavior
When postgresql database goes down the authentication plugin vmq_diversity_postgres.erl should try to reconnect till it succeed or the whole vernemq should shutdown because if the postgresql is not available, clients are not able to connect.

### Actual behaviour
It seems - from the log that vmq_diversity_postgres.erl plugin tries to reconnect several times, but then
give up on the connection and runs as if nothing happened and from that moment clients are not able to connect. We have environment with chaos monkey enabled and when it kills the postgresql, the vmq_diversity_postgres.erl doesn't reconnect.

Feb 13 16:56:58 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:58.014 [error] <0.446.0> CRASH REPORT Process <0.446.0> with 0 neighbours exited with reason: {empty,[{queue,get,[{[],[]}],[{fil
e,"queue.erl"},{line,187}]},{epgsql_sock,command_tag,1,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/epgsql/src/epgsql_sock.erl"},{line,409}]},{epgsql_sock,on_message,2,[{file,"/opt/
vernemq/distdir/1.3.0/_build/default/lib/epgsql/src/epgsql_sock.erl"},{line,684}]},{epgsql_sock,loop,1,[{file,"/opt/vernemq/distdir/1.3.0/_build/default/lib/epgsql/src/epgsql_sock.erl"},{lin
e,334}]},{gen_server,try_dispatch,4,[{file,"gen_server.erl"},{line,601}]},{gen_server,...},...]} in gen_server:terminate/7 line 812
Feb 13 16:56:58 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:58.512 [warning] <0.438.0>@vmq_diversity_worker_wrapper:handle_info:163 Could not connect to postgresql due to econnrefused
Feb 13 16:56:58 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:58.513 [error] <0.6689.0> gen_server <0.6689.0> terminated with reason: econnrefused
Feb 13 16:56:58 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:58.513 [error] <0.6689.0> CRASH REPORT Process <0.6689.0> with 0 neighbours exited with reason: econnrefused in gen_server:termi
nate/7 line 812

Feb 13 16:56:59 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:59.015 [warning] <0.431.0>@vmq_diversity_worker_wrapper:handle_info:163 Could not connect to postgresql due to econnrefused
Feb 13 16:56:59 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:59.015 [error] <0.6698.0> gen_server <0.6698.0> terminated with reason: econnrefused
Feb 13 16:56:59 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:59.015 [error] <0.6698.0> CRASH REPORT Process <0.6698.0> with 0 neighbours exited with reason: econnrefused in gen_server:terminate/7 line 812

Feb 13 16:56:59 mujsignal-220 vernemq[1137]: 2018-02-13 16:56:59.473 [error] <0.457.0>@vmq_diversity_script_state:handle_info:175 can't call function auth_on_register with args [{addr,<<"10.230.71.39">>},{port,56030},{mountpoint,<<>>},{client_id,<<"virtual-box-10">>},{username,<<"test:00:00:00:00:10">>},{password,nil},{clean_session,true}] in "/usr/share/vernemq/lua/auth/postgres.lua" due to {throw,{error,not_connected}}

Feb 13 16:57:43 mujsignal-220 vernemq[1137]: 2018-02-13 16:57:41.545 [error] <0.460.0> gen_server <0.460.0> terminated with reason: no function clause matching vmq_diversity_script_state:handle_info({#Ref<0.0.2.135979>,{error,not_connected}}, {state,10,{luerl,{array,26,100,undefined,{{{table,{array,0,10,nil,10},{{{{{empty,<<"_G">>,...},...},...},...},...},...},...},...}},...},...}) line 162

Feb 13 16:57:54 mujsignal-220 vernemq[1137]: 2018-02-13 16:57:54.465 [error] <0.7451.0>@vmq_mqtt_fsm:check_user:545 can't authenticate client {[],<<"virtual-box-04">>} due to
Feb 13 16:57:54 mujsignal-220 vernemq[1137]:                                 no_matching_hook_found

**Comment by larshesel** [comment]
Sounds like a bug. We'll look into it.

**Comment by yshadow** [comment]
Maybe if you support some http authentication plugin (like mosquitto) I can handle it myself.

**Comment by larshesel** [comment]
We support authentication via webhooks (http://vernemq.com/docs/plugindevelopment/webhookplugins.html) if that's what you mean?

**Comment by yshadow** [comment]
Yes that looks great. I just need to check if the common name from the client certificate is sent as username. If so than I have a solution.

**Comment by larshesel** [comment]
great - we still need to look at this bug at some point, though.

**Comment by yshadow** [comment]
Thank you Lars.

**Comment by dergraf** [comment]
I investigated this issue a bit today and as it seems that the only really annoying thing is the big amount of log output during the time the Postgres database is down. But besides that it seems VerneMQ recovers as expected when the database connection returns. 

It's by design that VerneMQ blocks new clients in case of a database isn't reachable.

As it looks like OP found a solution to this problem I am closing this for now.

Note:
I've also tested it with the newest version of epgsql (4.1.0, we're on 3.1.0), but it seems the similar issue exist.
